[STREAM-600] Add Video conferencing to the demo app

patrickfeeney03 opened the pull request:
- Join conferences
- Join meetings
- Mute video mute audio and end call
- Share screen

patrickfeeney03 committed:
get video working

patrickfeeney03 committed:
Add buttons to mute video and audio for Video in the Active Sessions table.
Add listeners to update video conference session.state and session.connectionState.

patrickfeeney03 committed:
Toggle audio properly depending on if its video or audio.

patrickfeeney03 committed:
Use the event sessionState change to figure out if a collaborateVideo call should be removed from the active sessions table.

patrickfeeney03 committed:
Let the accepSession method create its own media stream instead of creating our own.

patrickfeeney03 committed:
Add function to call the sdk function to join using meeting id also.

patrickfeeney03 committed:
Display room jid

patrickfeeney03 committed:
Give Video its own Active Sessions table. In progress.

patrickfeeney03 committed:
Revert "Give Video its own Active Sessions table. In progress."

This reverts commit a495d4e4a902bdde1139844dc5ff4eeb6fa33723.

patrickfeeney03 committed:
Revert "Revert "Give Video its own Active Sessions table. In progress.""

This reverts commit 5a468c880f8878b686828d24c22e4cd2eda13b5d.

patrickfeeney03 committed:
Set ActiveConversationsTable.tsx back to its original state

patrickfeeney03 committed:
Remove old sliced in conversationsSlice.
Convert styles to classes and keep them in Video.css for Video.tsx
Hide video when disabled.

patrickfeeney03 committed:
Call startMedia and pass the returned stream to acceptSession

patrickfeeney03 committed:
changelog

maxwellmooney13 commented on react-demo-app/src/components/Video.tsx:
```diff
@@ -0,0 +1,163 @@
+              accent="primary"
+              onClick={startVideoConf}
+            >
+              Place Conference
```
NIT: I think we should reword these buttons. `Join Conference` perhaps?

patrickfeeney03 commented on react-demo-app/src/components/Video.tsx:
```diff
@@ -0,0 +1,163 @@
+              accent="primary"
+              onClick={startVideoMeeting}
+            >
+              Place Meeting
```
This one isn't 1 to 1. I think this is group calls

patrickfeeney03 committed:
Move useEffect to the top of the file Video.tsx
Remove imports conversationsSlice.ts

patrickfeeney03 commented on react-demo-app/src/components/ActiveVideoConversationsTable.tsx:
```diff
@@ -0,0 +1,125 @@
+        }
+      ));
+    }
+  }, [videoConversations]);
```
@zservies What is your view on this? I added `getParticipantUsingDemoApp` to the dep array but it was causing infinite re renders; its a function. I think I could memoise the function using useCallback but that feels unnecessary.

patrickfeeney03 committed:
Undo comment

patrickfeeney03 committed:
Change wording for Join buttons DemoApp

patrickfeeney03 committed:
Removed the useEffect from ActiveVideoConversationsTable.tsx

patrickfeeney03 committed:
Store the loading states in each videoConversation rather than a global shared object

patrickfeeney03 committed:
Remove unused code in ActiveVideoConversationsTable.tsx

patrickfeeney03 committed:
* Hide vanity video right after it starts loading to prevent black screen
* Rename some things in conversationsSlice.ts

patrickfeeney03 committed:
* Use the promise from the sdk to remove the spinner instead of waiting for the participantsUpdate

zservies commented:
There doesn't seem to be a preview of the shared screen when screensharing, can this be added?

zservies commented on react-demo-app/src/components/Video.tsx:
```diff
@@ -0,0 +1,184 @@
+    <>
+      <Card className="video-card">
+        <h2>Video</h2>
+        <div className="buttons-and-video-container">
```
Can we extract this into a card and copy similar styling as the "Place a Call" card for softphone?

zservies commented on react-demo-app/src/components/Video.tsx:
```diff
@@ -0,0 +1,184 @@
+              </div>
+            </div>
+            <div>
+              <p>Yours</p>
```
Rename this "Local Video"

zservies commented on react-demo-app/src/components/Video.tsx:
```diff
@@ -0,0 +1,184 @@
+              <p>Yours</p>
+              <div className='video-container'>
+                <video ref={vanityVideoRef} autoPlay playsInline
+                       style={{visibility: outgoingStreamIsActive && !(
```
Can some of this be put into a function instead? I don't love having a bunch of logic in the template if we can prevent it, I think its better defined above.

zservies commented on react-demo-app/src/components/Video.tsx:
```diff
@@ -0,0 +1,184 @@
+            </div>
+            <div>
+              <p>Yours</p>
+              <div className='video-container'>
```
Can we also add these to cards as well?

patrickfeeney03 committed:
* Change `any`s to `unknown`s.
* Add strict mode
* Add video functions to useSdk
* Create VideoElements.tsx

patrickfeeney03 committed:
* Remove state for incoming and outgoing active streams

patrickfeeney03 committed:
* Active conference switcher v1

patrickfeeney03 committed:
* Select active video conv from table
* Remove sdk from Video.tsx and only use useSdk.

patrickfeeney03 committed:
* Make slice for Video conversations

patrickfeeney03 committed:
* Fix Video flickering

patrickfeeney03 committed:
* Some video styling

patrickfeeney03 committed:
* More styling

patrickfeeney03 committed:
* Move the Screen share button to activeConv table

patrickfeeney03 committed:
* show screen share

patrickfeeney03 committed:
* Mute audio in vanity video

patrickfeeney03 committed:
* Add better typing to useSdk.ts

patrickfeeney03 committed:
* remove a lot of generated css

patrickfeeney03 committed:
* Show errors related to joining video

patrickfeeney03 committed:
* Show errors related to joining video
* Working on CSS for table when meeeting id is too long

patrickfeeney03 committed:
* 'fix' the table. Fixed size now

patrickfeeney03 committed:
* Fix overflow in table

patrickfeeney03 committed:
* Remove most inline styles

patrickfeeney03 committed:
Merge branch 'develop' into demoApp-video

patrickfeeney03 committed:
* fix undefined error

patrickfeeney03 committed:
* oops undo edit to ActiveConversationsTable.css

patrickfeeney03 committed:
* Create css file for ActiveVideoConversationsTable.tsx

patrickfeeney03 committed:
* Add svg when active speaker has camera off

patrickfeeney03 committed:
* Remove useEffect from VideoElements.tsx. Apparently I did not actually need it.

patrickfeeney03 committed:
* Store active participants in each conversation in slice

patrickfeeney03 committed:
Merge branch 'develop' into demoApp-video

patrickfeeney03 committed:
* Add border to video when 'speaking'

patrickfeeney03 committed:
Merge branch 'develop' into demoApp-video

patrickfeeney03 committed:
* broken package-lock.json?

patrickfeeney03 committed:
* remove unsued background-color VideoElements
* Add userId at bottom

patrickfeeney03 committed:
* Set default value for devices on startup
* Join conference by pressing Return key

patrickfeeney03 committed:
* Show 'waiting for others to connect...'

patrickfeeney03 committed:
* remove unsued parameter in a function

patrickfeeney03 committed:
* break Video.tsx into smaller files

patrickfeeney03 committed:
* doing some refactoring bc components are too big

patrickfeeney03 committed:
* transfer logic to VideoElement.tsx from VideoElements.tsx

patrickfeeney03 committed:
Merge branch 'develop' into demoApp-video

# Conflicts:
#	react-demo-app/package-lock.json

patrickfeeney03 committed:
* Add typing to slice

patrickfeeney03 committed:
* typing

patrickfeeney03 committed:
* styling

patrickfeeney03 committed:
* typing

zservies commented on react-demo-app/src/components/PendingSessionsTable.tsx:
```diff
@@ -2,19 +2,20 @@ import './PendingSessionsTable.css';
   function handlePendingSession(accept: boolean, conversationId: string): void {
     if (accept) {
-      sdk.acceptPendingSession({ conversationId });
+      sdk?.acceptPendingSession({ conversationId });
```
Why the optional chaining here? I would think if we've auth'd we should have an SDK instance.

zservies replied:
Same goes in a lot of other places.

patrickfeeney03 replied:
that sdk comes from useSelector which could be undefined.  I will try to fix it in the sdkSlice so it stops thinking it could be null

zservies commented on react-demo-app/src/features/conversationsSlice.ts:
```diff
@@ -35,12 +32,12 @@ export const conversationsSlice = createSlice({
-        delete state.activeConversations[id];
-      }
-    },
+    // removeConversations: (state, action) => {
```
Uncomment.

patrickfeeney03 replied:
I think I was checking if it was required. It's never called.

zservies commented on react-demo-app/src/hooks/useSdk.ts:
```diff
@@ -76,22 +77,47 @@ export default function useSdk() {
+
   // If a pendingSession was cancelled or handled, we can remove it from our state.
-  function handleCancelPendingSession(pendingSession: IPendingSession): void {
+  function handleCancelPendingSession(pendingSession: ISessionIdAndConversationId): void {
```
Why'd this type change?

patrickfeeney03 replied:
Because TypeScript was complaining that IPendingSession was not the right one.

<img width="894" height="231" alt="image" src="https://github.com/user-attachments/assets/8dc5f6ee-4b1e-48d8-acbf-bce13dc4ae5e" />

After I added typing to the SDK I saw type errors like that one so fixed it.

The types come from the SDK itself, I did not manually make them. I just told the useSelector what type everything was.
<img width="778" height="58" alt="image" src="https://github.com/user-attachments/assets/f92ae74b-4895-4a9d-8cd1-7e1c630d8d42" />

zservies commented on react-demo-app/src/hooks/useSdk.ts:
```diff
@@ -107,15 +133,18 @@ export default function useSdk() {
   }
-  function handleGumRequest(state: void): void {
-    dispatch(updateGumRequests(state));
+  function handleGumRequest(_state: ISdkGumRequest): void {
```
Why'd this change?

zservies commented on react-demo-app/src/hooks/useSdk.ts:
```diff
@@ -127,10 +156,10 @@ export default function useSdk() {
     sdk.media.enumerateDevices(true);
   }
-  function requestDevicePermissions(type: string): void {
+  function requestDevicePermissions(type: 'audio' | 'video' | 'both'): void {
```
I think this should just be a string.

patrickfeeney03 replied:
The sdk has that type and will complain if its string

zservies commented on react-demo-app/src/components/Video/VideoElement.tsx:
```diff
@@ -0,0 +1,64 @@
+  videoRef: RefObject<HTMLVideoElement>,
+  userId: string | undefined
+}) {
+  const SVG_LINK = "https://dhqbrvplips7x.cloudfront.net/volt/1.12.1-178/assets/default-person.svg"
```
is this not coming from spark? We shouldn't be pointing to a Volt cloudfront URL here for security.

patrickfeeney03 replied:
I'll change it. I found a similar one in Spark.

patrickfeeney03 commented on react-demo-app/src/hooks/useSdk.ts:
```diff
@@ -55,7 +56,7 @@ export default function useSdk() {
     webrtcSdk.on('cancelPendingSession', handleCancelPendingSession);
     webrtcSdk.on('handledPendingSession', handledPendingSession);
-    webrtcSdk.on('sessionStarted', handleSessionStarted);
+    // webrtcSdk.on('sessionStarted', handleSessionStarted);
```
I set it up in a function below. I wanted to run different callbacks since there could be different conversations going on.

patrickfeeney03 commented on react-demo-app/src/hooks/useSdk.ts:
```diff
@@ -76,22 +77,47 @@ export default function useSdk() {
     sdk.startSoftphoneSession({ phoneNumber });
   }
 
+  function sessionStarted(callback: (arg0: VideoMediaSession) => Promise<void>) {
```
Maybe I'm overcomplicating things here.

I was trying to make a function to set up the listener and also pass in the function to call when the listener triggered.
So `callback` is the handler in Video.tsx and `arg0` is the session that sessionStarted provides. So I'd pass that `arg0` to the callback that was passed.

I'll try to simplify this

patrickfeeney03 committed:
remove functions that are no longer needed

remove useVideoSession.ts!
Moved the event listener to sessionStarted to useSdk part1
resolve some pr comments

patrickfeeney03 committed:
Merge branch 'develop' into demoApp-video

patrickfeeney03 committed:
Remove useVideoForm.ts

patrickfeeney03 committed:
No longer store session in Redux. Pull it from SDK store.

patrickfeeney03 committed:
remove repeated code and console log

patrickfeeney03 committed:
remove repeated code and console log

patrickfeeney03 committed:
Merge remote-tracking branch 'origin/demoApp-video' into demoApp-video

zservies commented on react-demo-app/src/components/Video/ActiveVideoConversationsTable.tsx:
```diff
@@ -0,0 +1,181 @@
+    dispatch(setCurrentlyDisplayedConversation({ conversationId }));
+  };
+
+  const isLocalPartSharingScreen: boolean[] =
```
NIT: just use the longform so its more clear - `isLocalParticipantScreensharing`

zservies commented on react-demo-app/src/components/Video/VideoElements.tsx:
```diff
@@ -0,0 +1,85 @@
+import VideoElement from "./VideoElement.tsx";
+import useSdk from "../../hooks/useSdk.ts";
+
+export default function VideoElements({
```
Cleanup indentation

zservies commented on react-demo-app/src/components/ActiveConversationsTable.tsx:
```diff
@@ -4,20 +4,21 @@ import './ActiveConversationsTable.css';
     if (conversations.length) {
-      setHoldLabels(conversations.map((convo) => convo.mostRecentCallState.held ? 'Unhold' : 'Hold'));
-      setMuteLabels(conversations.map((convo) => convo.mostRecentCallState.muted ? 'Unmute' : 'Mute'));
+      setHoldLabels(conversations.map((convo) => convo?.mostRecentCallState!.held ? 'Unhold' : 'Hold'));
```
Should these be optional?

patrickfeeney03 committed:
temp

patrickfeeney03 committed:
We don't need to keep memberStatusUpdate as a useRef

patrickfeeney03 committed:
comment

patrickfeeney03 committed:
remove unused logging

patrickfeeney03 committed:
Mute VideoElement if it comes from local

patrickfeeney03 committed:
Add memberStatusUpdate to video.md docs

patrickfeeney03 committed:
Simplify on screen user and speaker detection

patrickfeeney03 committed:
Update docs about Guests not being a thing

patrickfeeney03 committed:
Address PR comments

patrickfeeney03 committed:
Merge branch 'develop' into demoApp-video

# Conflicts:
#	react-demo-app/package-lock.json
#	react-demo-app/src/components/ActiveConversationsTable.tsx
#	react-demo-app/src/components/AudioDevices.tsx
#	react-demo-app/src/components/HandledPendingSessionsTable.tsx
#	react-demo-app/src/components/MediaStateCounts.tsx
#	react-demo-app/src/components/OutputDevices.tsx
#	react-demo-app/src/components/PendingSessionsTable.tsx
#	react-demo-app/src/components/Softphone.tsx
#	react-demo-app/src/components/StationDetails.tsx
#	react-demo-app/src/components/VideoDevices.tsx
#	react-demo-app/src/features/sdkSlice.ts
#	react-demo-app/src/hooks/useSdk.ts

patrickfeeney03 committed:
Fixing merge conflicts errors

patrickfeeney03 committed:
Work on some Linting errors

patrickfeeney03 committed:
Get UI and SDK devices in-sync and save preferences in LocalStorage.

Use LocalStorage instead of the equivalent backend service used in Volt and webdir.

patrickfeeney03 committed:
Revert "* Add typing to slice"

This reverts commit 5bf34244

Revert "* styling"

This reverts commit be7424d8

patrickfeeney03 committed:
Undo some changes to reduce diff. Fix types

Losing it. Default device stuff. Also undo changes to conversationsSlice.ts

Type

patrickfeeney03 committed:
- Move the slider from Audio to Output device since it changes the output volume not the mic's.
- Query audio and video elements using DOM directly from the useSdk hook

patrickfeeney03 committed:
Fix video clipping through div by hiding it.

patrickfeeney03 committed:
- Don't reassing if same values
- Rename interface.

patrickfeeney03 committed:
styling

patrickfeeney03 committed:
fix diff again

patrickfeeney03 committed:
Remove mute

patrickfeeney03 committed:
Initialize conv.usersTalking

patrickfeeney03 committed:
Remove unused code
